Update strings in UI without reloading (fixes translation on Android!)

lucyhe opened the pull request:
Problem in a nutshell:
- The only way to force Polymer bindings to update is by updating a function parameter in the bound Polymer expression.
- Before this change, we mostly had constants in the translation Polymer expressions, e.g. {{ "APP_MISSING_TITLE" | $$ }}
- Therefore, changing languages required saving the new language to local storage, and then refreshing the UI
- If loading from global storage took too long, the UI would get stuck in a reload loop: first, the default language (English) would load, another language French (for example) is loaded from storage, the UI reloads, but then English is used again while waiting for storage to load...etc

Solution:
- Add model.globalSettings.language to each translation Polymer expression, forcing it to update if the language updates
- Remove reloading code related to i18n
- important Remove all calls to ui.i18n_t made at start up time. This is because there are no observers on the input or output of ui.i18n_t and we can not refresh strings created by ui.i18n_t when the language has changed. Instead, ui.i18n_t should only be used when we are adding new text to the DOM (after start up), for example, in a dialog. This will work because when calls to ui.i18n_t are made on the fly, the translator module will be able to refer to the most recently set language.
- Example: connectedNetworks was a string in settings.ts/settings.html that would get translated with ui.i18n_t at start up. It was referred to in Polymer by {{ connectedNetworks }} <-- without a reference to the globalSettings.language, this string will not know to update if the user changes the language! Instead we need to move the logic for setting connectedNetworks into Polymer translation expressions.

**Most important things (will add to docs):**
Translation in HTML files before: {{ "PLACEHOLDER" | $$({param1: value1...}) }}
Translation in HTML files after:{{ "PLACEHOLDER" | $$(model.globalSettings.language, {param1: value1...}) }}
Translation in JS files before: translator.i18n_t("PLACEHOLDER", {param1: value1...})
Translation in JS files after: ui.i18n_t("PLACEHOLDER", {param1: value1...}) <-- same but you cant do this for any strings needed at start up time.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2705)
<!-- Reviewable:end -->

dborkan commented:
I had a crazy idea: can we re-assign the $$ function each time the language changes, so we don't have to do all this replacing of ```$$``` with ```$$(model.globalSettings.language)```?  My understanding of that part of the pull request is that without any changes, an expression like ```{{ "APP_MISSING_TITLE" | $$ }}``` won't get re-evaluated if the language changes, because nothing in that expression changes, but if we re-assign $$ a new function that translates to the newly selected language, then maybe it will get re-evaluated?

lucyhe commented:
Chatted in person, but for reference:

I couldn't get this to work in uProxy, so I tried out the basic concept in a Polymer fiddle: https://jsfiddle.net/uq238zu7/2/. It seems like changing the function referenced by the filter will not cause the bound expression to refresh/update. This can be confirmed in the fiddle because calling the filter function (via the "alert" button) can produce a different result than the string bound in the template (which never updates).

I suspect that updating the filter function will not update the bound expressions because Polymer elements have no "watcher" for a global Polymer filter. Because of this, we tried two other approaches:

1) Use {{ ui.i18n_t("MESSAGE_STR") }} instead of {{ "MESSAGE_STR" | $$ }}, since this.ui *should* be watched for updates. Unfortunately this produced a `Only identifier function invocations are allowed` error; quick search showed that Polymer doesn't allow us to call functions of other objects in expressions, but only functions defined in that element (i.e. any foo for this.foo).

2) Given the insight learned from 1), I added a new (local) function, $$, to the Polymer element. this.$$ is set to ui.i18n_t initially, and then when the language changed, we updated the pointer of this.$$ to a new i18n_t of the new language. Unfortunately this also didn't work. I again tried the basic concept in a Polymer fiddle: https://jsfiddle.net/jtmu5v3u/1/

I think Polymer just must not watch for function reference changes at all, or at least they are disregarded wrt when to update bound expressions. The [expression documentation](https://docs-05-dot-polymer-project.appspot.com/0.5/docs/polymer/expressions.html#expression-syntax) does only mention that refreshes happen when parameters update: "The function’s arguments are observed for changes, and the expression is re-evaluated whenever one of the arguments changes."

jpevarnek commented:
Reviewed 3 of 36 files at r1, 28 of 35 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/generic_ui/polymer/advanced-settings.ts, line 88 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2705#-KPTHsqM1u0eSVnHczhv:-KPTHsqM1u0eSVnHczhw:b-gcxayr) ([raw file](https://github.com/uproxy/uproxy/blob/fa1ecfe24b63aeb83efc1e590c2b248919b71952/src/generic_ui/polymer/advanced-settings.ts#L88)):*
> ```XML
>     });
>   },
>   created: function() {
> ```

Any reason for ready vs. created?

---

*[src/generic_ui/polymer/description.html, line 48 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2705#-KPTIBeuGbLjGnuSbrD0:-KPTIBeuGbLjGnuSbrD1:b-lx8zc5) ([raw file](https://github.com/uproxy/uproxy/blob/fa1ecfe24b63aeb83efc1e590c2b248919b71952/src/generic_ui/polymer/description.html#L48)):*
> ```HTML
>     </div>
>     <div id='savedDescription' hidden?='{{ editing }}' on-tap='{{ editDescription }}'>
>       <span hidden?='{{!model.globalSettings.description}}'>{{model.globalSettings.description}}</span>
> ```

Can we delete the function from `description.ts`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2705)*
<!-- Sent from Reviewable.io -->

lucyhe commented:
Review status: 30 of 32 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/generic_ui/polymer/advanced-settings.ts, line 88 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2705#-KPTHsqM1u0eSVnHczhv:-KPT_cUxMi_SseD2xbDn:b-3cxbuf) ([raw file](https://github.com/uproxy/uproxy/blob/fa1ecfe24b63aeb83efc1e590c2b248919b71952/src/generic_ui/polymer/advanced-settings.ts#L88)):*
<details><summary><i>Previously, jpevarnek (Jonathan Pevarnek) wrote…</i></summary>
> Any reason for ready vs. created?
</details>
Nope, changed.

---

*[src/generic_ui/polymer/description.html, line 48 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2705#-KPTIBeuGbLjGnuSbrD0:-KPT_es2QYFyRq9QQ8Wt:b-ubb87t) ([raw file](https://github.com/uproxy/uproxy/blob/fa1ecfe24b63aeb83efc1e590c2b248919b71952/src/generic_ui/polymer/description.html#L48)):*
<details><summary><i>Previously, jpevarnek (Jonathan Pevarnek) wrote…</i></summary>
> Can we delete the function from `description.ts`?
</details>
Good catch, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2705)*
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/generic_ui/polymer/advanced-settings.ts, line 88 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy/2705#-KPTHsqM1u0eSVnHczhv:-KPTcTRxqWZdmqGd6tX6:b9evp5j) ([raw file](https://github.com/uproxy/uproxy/blob/fa1ecfe24b63aeb83efc1e590c2b248919b71952/src/generic_ui/polymer/advanced-settings.ts#L88)):*
<details><summary><i>Previously, lucyhe wrote…</i></summary>
> Nope, changed.
</details>
I think there's a few other places we used `created`, I have no problem with it if it does something for us, but otherwise it would be nice to keep the code consistent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2705)*
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2705)*
<!-- Sent from Reviewable.io -->

jpevarnek commented:
Reviewed 7 of 7 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/generic_ui/polymer/advanced-settings.html, line 222 \[r5\]](https://reviewable.io:443/reviews/uproxy/uproxy/2705#-KQMGNAsqXb-5o0TwSn5:-KQMGNAsqXb-5o0TwSn6:b8xpj8c) ([raw file](https://github.com/uproxy/uproxy/blob/422a19565fbd89d0b1aad2495bbd4a86aef7fd50/src/generic_ui/polymer/advanced-settings.html#L222)):*
> ```HTML
>               </span>
>               <span class='advancedSettingsText' hidden?='{{ _reproxyCheckToString(reproxyCheck) !== "TRUE" }}'>
>                 {{ 'TOR_PORT_TRUE' | $$(model.globalSettings.language) }} {{ testedTorPort  | $$(model.globalSettings.language) }}
> ```

Isn't `testedTorPort` a number?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2705)*
<!-- Sent from Reviewable.io -->

lucyhe commented:
Fixed, the original string should have had a placeholder instead of the port number after an i18n-ed string.

Thanks!

jpevarnek commented:
:thumbsup:

---

Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2705#-:-KQMc-eWSfxgpNojJi-R:bn1wz0t)*
<!-- Sent from Reviewable.io -->
